### ISSUE TITLE
Libyang 2.1.128 has breaking API change compared to 2.1.80

### ISSUE THREAD

**Issue by mwinter-osr** [CESNET/libyang#2152]
It looks like the new release 2.1.128 has a breaking API change and is NOT compatible with 2.1.80
Based on the minor version change, this MUST NOT have happened.

The new 2.1.128 is missing a symbol  'lyd_parent'

This was found with FRR, which fails to compile 8.x versions with libyang 2.1.128, but works fine with 2.1.80

**Comment by jktjkt** [comment]
That sounds strange because 2.1.28 is the first tag which contains commit d3b351a209a9d03bc69df647514f16525e48b0a6. That commit actually *introduced* these symbols to the public library. Previously, these were defined as `static inline` functions in the public headers, meaning that they were not present as symbols in the library itself.

Can you please describe what exactly you're doing, and what failure and when you're getting?

**Comment by idryzhov** [comment]
In 2.1.80, `lyd_parent` was a `static inline` function defined in a header. Some FRR daemons were using this function by including the header without linking with `libyang` and it was working. In 2.1.128, `lyd_parent` become a regular function, so now daemons need to be linked with `libyang` to use this function, otherwise build fails.

I'm not sure we can say that it's a breaking API change, because the API was not changed. Maybe it's an ABI breaking change, I don't know. Anyway, I suppose it's not really safe to use some library functions without actually linking with this library :) We fixed this on FRR side by adding necessary linking.

**Comment by michalvasko** [comment]
I see, good that you managed to fix it without problems. But it seems the reason for this change was fairly weak and if I confirm that the compiler is not able to optimize the functions as well as when they are inlined, I will revert it for the next major SO version release, which will most likely be the next release (but no worries, the changes will all the rather small and in almost all cases require only recompilation).

**Comment by idryzhov** [comment]
Well, reverting it back will be a breaking change for sure, because all binaries compiled with 2.1.128 will stop working because of the missing symbol. So, please, don't do the revert.

**Comment by michalvasko** [comment]
Yes, that is why I want to increase the major SO version as well, would that be a problem for you? I have tested it and it seems the inline functions are significantly faster so I would prefer to revert it.

**Comment by irfanHaslanded** [comment]
Since things were working before as inline functions for you @idryzhov, IMHO, it shouldn't be a problem to revert?

**Comment by choppsv1** [comment]
> Since things were working before as inline functions for you @idryzhov, IMHO, it shouldn't be a problem to revert?

That's not the issue, the issue is that when this symbol disapears from the library then all the compiled binaries that count on it being there are going to break.

**Comment by idryzhov** [comment]
@michalvasko generally I think this is ok to do such a change in a major release. But I suppose some care is needed from deb/rpm package maintainers when upgrading.

**Comment by choppsv1** [comment]
> Yes, that is why I want to increase the major SO version as well, would that be a problem for you? I have tested it and it seems the inline functions are significantly faster so I would prefer to revert it.

How are you going to keep this from breaking the minute someone `apt-get` upgrades their libyang to your new version, and the older working SO is removed from their system? You have to change the project major as well.

**Comment by choppsv1** [comment]
You need to think about a different solution here, IMO.

**Comment by choppsv1** [comment]
I think maybe you should try and get the symbol included with external linkage in the SO (e.g., compile w/o the static inline present), and also include the static inline in the header file.

**Comment by michalvasko** [comment]
@choppsv1 The packages are usually named "<package-name><major-SO-version>" and in those cases upgrading should not be a problem. But I have just checked and the SO version is not included on all the distributions, it is up to those maintainers to solve these issues, then. I will add the version to the package scripts included in our projects (it was for DEB but not for RPM).

**Comment by choppsv1** [comment]
> @choppsv1 The packages are usually named `<package-name><major-SO-version>` and in those cases upgrading should not be a problem. But I have just checked and the SO version is not included on all the distributions, it is up to those maintainers to solve these issues, then. I will add the version to the package scripts included in our projects (it was for DEB but not for RPM).

Please don't do this, just ship it with both the external linkage version and the static inline. Everybody is happy and nothing breaks and no version foo is required.

**Comment by michalvasko** [comment]
I would not be that happy because we would need to maintain some hacks to achieve what you are asking for. There will be other NBC changes (mostly in logging it seems) so this one should not matter anymore. We wanted to release a new SO version not because of this specific problem, we can just include it as one of the changes since we want to make the release anyway.

**Comment by choppsv1** [comment]
> I would not be that happy because we would need to maintain some hacks to achieve what you are asking for. There will be other NBC changes (mostly in logging it seems) so this one should not matter anymore. We wanted to release a new SO version not because of this specific problem, we can just include it as one of the changes since we want to make the release anyway.

I do not understand how just changing the SO version is going to cover the package upgrade. I think you're going to create a bunch of new headaches for us (and your users).

**Comment by choppsv1** [comment]
Ok well those are more major changes and I agree would warrant the change. Will development be ending on libyang2 then?

**Comment by michalvasko** [comment]
Well, yes, I see no reason someone would want to stay with the current version. The effort required with updating to v3 would be similar to any other *libyang* release, the difference being only the minor NBC changes.
